src/content/blog/first-four-val-town-runtimes.mdx

```diff
@@ -0,0 +1,247 @@
+The easiest way to run an arbitrary string of code is with the `eval()` method,
+which doesn't provide any kind of sandbox protections, so it can easily become
+an attack vector.
+
+Security generally means isolation and containment:
+
+- Vals need to be isolated from the system they're running on: they shouldn't
+  be able to read from the filesystem or learn about environment variables.
+- Vals should be isolated from each other: you shouldn't be able to affect
+  anyone else's val's behavior.
+
+#### Speed
+
+On the other hand, if you were to make every Val run in a full-fledged Docker
+container, it'd be tremendously slow. You're booting up a whole Linux machine
+with all the bells and whistles just to evaluate a few lines of code. It'd be
+[pretty secure](https://docs.docker.com/engine/security/#conclusions), but
+pretty slow.
+You could create a pool of Docker containers, but the memory and devops overhead
+would be significant.
+
+#### Flexibility
+
+Val Town lets you bring a lot of experience from other systems. If you've used
+Node.js, great - your NPM modules should mostly work. Thanks to our current system,
+a lot of DOM methods work too - you can parse query strings with [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) and use new APIs
+like web-standard streams. A way to square the safety + speed requirements
+would be to use [QuickJS](https://bellard.org/quickjs/), a standalone JavaScript
+engine that's fast to start up and easily sandboxed with WebAssembly. But
+we'd be miles away from NPM or Node compatibility, and would run JavaScript
+about 35x slower.
+
+---
+
+So, those are the properties we're aiming for. What have we tried so far? Buckle up.
+
+#### Runtime 1: Node.js + vm
+
+The very first version of Val Town used Node.js’s [vm module](https://nodejs.org/api/vm.html).
+Schematically, running a val looked like this:
+
+```ts
+const [returnValue, logs, error] = vm.runInContext(`
+  let resolveValue;
+  let returnValue;
+  let error;
+  const logs = [];
+  console.log = (log) => { logs.push(log); return undefined };
+  try {
+    ${replacedCode}
+    returnValue = ${name}
+  } catch (e) {
+    error = e;
+  }
+  [returnValue, logs, error];
+`);
+```
+
+Now, the documentation for the vm module warns that "The node:vm module
+is not a security mechanism. Do not use it to run untrusted code."
+So it should be no surprise that this approach didn't last very long.
+
+The Node.js vm module is incredibly convenient to use, but it was dangerous and
+didn't check our boxes:
+
+- Infinite loops in user code will take down your server
+- It's easy to escape the sandbox and access sensitive environment variables
+- You can't install NPM modules within the sandbox
+- Scripts that run in the Node.js VM sandbox run in _script_ mode, not _module_
+  mode, so you can't use `import` or `export`.
+
+#### Runtime 2: Node.js + vm2
```

Comment:
Again, you could caveat that we moved off this a month after we launched, or something like that